### PR TITLE
Extension of transform module

### PR DESCRIPTION
**Pull request by EnricoTrizio**
## Description
Adding some preprocessing tools to the **transform module**, including descriptors computation and tools.

**NB-1** Transform object are supposed not to have trainable parameters.
**NB-2** Position-dependent descriptors are implemented only for cubic/orthorhombic cells.

The structure of the module is:
### Tools
- Normalization
- Switching functions (new)
- Continuous Histogram (new, derivable histogram using KDE)

### Descriptors, acting on positions
- Multiple descriptors (new, wrapper class to compute multiple descriptors on the same set of positions)
- Pairwise distances (new, computes ALL pairwise distances from a set of positions)
- Torsional angle (new, computes the torsional angle from a set of positions + index of 4 involved atoms)
- Coordination number (new, computes from a set of positions coordination numbers of elements of group A wrt group B)
- Eigs adjacency matrix (new, computes adjacency matrix eigenvalues. Could be moved to a tutorial)

#### Already implemented but likely to be kept out of this PR:
- Radius graph (compute graph inputs in the format: edge sources, edge destinations and edge distances)
- Radial distribution function (compute RDF from a set of position and optionally the integral of a peak)
 
## Status
- [x] Ready to go

### CODE REVIEW

File: mlcolvar/core/transform/utils.py [hunk 1]
```diff
@@ -2,83 +2,7 @@
...
-
-    def __repr__(self):
-        repr = "<Statistics>  "
-        for prop in self.properties:
-            repr += f"{prop}: {getattr(self,prop).numpy()} "
-        return repr
+__all__ = ["Inverse"]
 
 
 class Inverse(torch.nn.Module):
```
Comment (luigibonati) [inline mlcolvar/core/transform/utils.py:8]: let's add also a function to stack multiple transforms (e.g. to combine distances and angles)

File: mlcolvar/core/transform/descriptors/pairwise_distances.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    """
+    Pairwise distances transform, compute all the non duplicated pairwise distances for a set of atoms from their positions
+    """
+
+    def __init__(self, 
+                 n_atoms : int,
+                 PBC: bool,
+                 cell: Union[float, list],
+                 scaled_coords : bool = False,
+                 slicing_indeces : list = None) -> torch.Tensor:
```
Comment (luigibonati) [inline mlcolvar/core/transform/descriptors/pairwise_distances.py:20]: we could slice based on atom pairs like it is done in mdtraj 
https://www.mdtraj.org/1.9.8.dev0/api/generated/mdtraj.compute_distances.html#mdtraj.compute_distances

File: mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        self.cutoff = cutoff 
+        self.n_atoms = n_atoms
+        self.PBC = PBC
+        self.cell = cell
+        self.scaled_coords = scaled_coords
+        self.switching_function = switching_function
+
+    def compute_adjacency_matrix(self, x):
+        x = compute_adjacency_matrix(pos=x,
+                                     mode = self.mode,
```
Comment (andrrizzi) [inline mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py:62]: I just know you'll hate me by the end (or beginning?) of the review, but PEP8 wants no space around ``=`` for function arguments.

File: mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                                     cutoff = self.cutoff, 
+                                     n_atoms = self.n_atoms,
+                                     PBC = self.PBC,
+                                     cell = self.cell,
+                                     scaled_coords = self.scaled_coords,
+                                     switching_function=self.switching_function)
+        # x = torch.squeeze(x)
```
Comment (andrrizzi) [inline mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py:69]: Another suspicious commented line to double check.

File: mlcolvar/core/transform/tools/switching_functions.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    # ========================== define here switching functions ==========================
+    def Fermi_switch(self,
+                     x : torch.Tensor, 
+                     cutoff : float, 
+                     q : float = 0.01, 
+                     prefactor_cutoff : float = 1.0):
+        y = torch.div( 1, ( 1 + torch.exp( torch.div((x - prefactor_cutoff*cutoff) , q ))))
+        return y
+
+    def Rational_switch(self,
```
Comment (andrrizzi) [inline mlcolvar/core/transform/tools/switching_functions.py:56]: Would it be a good idea to implement these as two separate classes with a common base class instead?

**Review by andrrizzi — COMMENTED** [review]
Looks good! I've left a few comment but it's only nitpicking.

For the structure, it looks ok to me, but I'd tend to have the least possible number of packages/modules called ``tools`` or ``utils``. I might be missing the rationale here though. I usually use tools essentially as synonym for utils meaning "anything I don't know where else to put/how else to organize".

Maybe we could simply move all the ``transform/tools/*.py`` modules directly in ``transform/*.py``. Some of the elements (like ``Inverse`` or ``Statistics``) might deserve their own ``inverse.py`` or ``statistics.py`` module. This is mostly taste anyway. It might not be the prettiest structure to have a bunch of small modules inside a package but I usually find it helpful when I navigate codebases.

File: mlcolvar/core/transform/descriptors/utils.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+        dist_components = dist_components - shifts
+
+    # if we used scaled coords we need to get back to real distances
+    if scaled_coords:
+        dist_components = torch.einsum('bijk,i->bijk', dist_components, cell)
+
+    return dist_components
+
+
+def compute_distances_matrix(pos : torch.Tensor,
```
[the remark this answers is not in the document]
Comment (EnricoTrizio) [reply mlcolvar/core/transform/descriptors/utils.py:142]: let's just keep a compute_distances_matrix function with the key vector=False for returning the components of the vectorial distances

File: mlcolvar/core/transform/descriptors/coordination_numbers.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        Returns
+        -------
+        torch.Tensor
+            Coordination numbers of elements of group A with respect to elements of group B
+        """
+        super().__init__(in_features=int(n_atoms*3), out_features=len(group_A))
+        
+        # parse args
+        self.group_A = group_A
+        self.group_A_size = len(group_A)
```
[the remark this answers is not in the document]
Comment (EnricoTrizio) [reply mlcolvar/core/transform/descriptors/coordination_numbers.py:60]: Do you think it is needed if the transforms are meant to be preprocessing tools? Maybe making them private could be enough?

File: mlcolvar/core/transform/descriptors/coordination_numbers.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        self.PBC = PBC
+        self.cell = cell
+        self.scaled_coords = scaled_coords
+
+        self.mode = mode
+        self.switching_function = switching_function
+    
+    def compute_coordination_number(self, pos):
+        # move the group A elements to first positions
+        # pos = pos[self.reordering]
```
[the remark this answers is not in the document]
Comment (EnricoTrizio) [reply mlcolvar/core/transform/descriptors/coordination_numbers.py:77]: Thanks for spotting it, now it should work and it is also covered in the tests. I added a new calculation in which I manually swapped two atoms and added a check on the results with respect to the original data.

File: mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py (hunk 4 above)
Comment (EnricoTrizio) [reply mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py:69]: It should be correct

File: mlcolvar/core/transform/utils.py (hunk 1 above)
Comment (EnricoTrizio) [reply mlcolvar/core/transform/utils.py:8]: Should we do it in general for transforms or for descriptors only?  Because the combined transforms should have the same input I think, e.g. the atomic positions of the same system

File: mlcolvar/core/transform/descriptors/pairwise_distances.py (hunk 2 above)
Comment (EnricoTrizio) [reply mlcolvar/core/transform/descriptors/pairwise_distances.py:20]: I changed the slicing from number of the distance to the atom pairs list

File: mlcolvar/core/transform/descriptors/coordination_numbers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrrizzi) [reply mlcolvar/core/transform/descriptors/coordination_numbers.py]: Yeah, making them private totally works!

File: mlcolvar/core/transform/tools/switching_functions.py (hunk 5 above)
Comment (andrrizzi) [reply mlcolvar/core/transform/tools/switching_functions.py:56]: If you mean so that you can pass a simple string to decide which one to use, yeah, it make a lot of sense. You can still do it using an object-oriented design using a factory-like pattern if you wanted. For just two classes, something like this could be enough:
```python
def get_switching_function(type_str):
    if type_str == 'Fermi':
        return FermiSwitchingFunction
    elif ...:
        ...
    else:
        raise ValueError
```
and then in the classes that need the switching function you just do
```python
get_switching_function('Fermi')(in_features, cutoff, options)
```
Factory patterns are often implemented with a registry rather than hardcoding with ``if-else``s, but for now I'd keep it as simple as possible.

Again, this is matter of taste so it's not essential. The current implementation works.

File: mlcolvar/core/transform/utils.py (hunk 1 above)
Comment (luigibonati) [reply mlcolvar/core/transform/utils.py:8]: yes I believe the stack should be used only for descriptors

Comment (EnricoTrizio) [reply mlcolvar/core/transform/utils.py:8]: perfect, it's in a separate multiple_descriptors module

File: mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py (hunk 3 above)
Comment (EnricoTrizio) [reply mlcolvar/core/transform/descriptors/eigs_adjacency_matrix.py:62]: Don't worry I hate you already, I'll fix this nonetheless

File: mlcolvar/core/transform/tools/switching_functions.py (hunk 5 above)
Comment (EnricoTrizio) [reply mlcolvar/core/transform/tools/switching_functions.py:56]: I already gave you the spaces before the double dots. Jokes apart, I also kept a structure similar to the way we pass the options to the CV models. So maybe we can keep it

File: mlcolvar/core/transform/descriptors/multiple_descriptors.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
+import torch
+
+__all__ = ["MultipleDescriptors"]
+
+class MultipleDescriptors(torch.nn.Module):
+    """Wrapper class to combine multiple descriptor transform objects acting on the same set of atomic positions in a single module"""
```
Comment (luigibonati) [inline mlcolvar/core/transform/descriptors/multiple_descriptors.py:6]: why should they act on the same atoms? e.g. a single torsional angle cannot be combined with pairwise distances on others?

Comment (EnricoTrizio) [reply mlcolvar/core/transform/descriptors/multiple_descriptors.py:6]: To have a single object the forward must be common, so does be the input. 
If you want to combine different descriptors you take as input all the positions and then you use the indexing within each single class to use only the atoms you need (as for torsional angle) or return only the output you want (as for the distances)
